### PR TITLE
Remove post-collation truncation from DPO

### PR DESCRIPTION
**Pull request by albertvillanova**
Remove post-collation truncation from DPO.

This PR removes internal truncation logic from the DPO trainer and require custom data collators to handle truncation themselves. This simplifies the trainer code and clarifies the contract for custom collators. Additionally, the PR updates documentation and error handling to reflect these changes.

Follow-up to:
- https://github.com/huggingface/trl/pull/5305#issuecomment-4083192554

### Motivation

Both built-in DPO collators (`DataCollatorForPreference` and `DataCollatorForVisionPreference`) already truncate sequences internally before padding. The only reason `_truncate_inputs` still existed in the trainer was as a silent safety net for custom collators, which is arguably worse than no safety net, because it hid the fact that the collator wasn't doing its job.

This PR makes the contract explicit and removes the silent fix-up.

### Changes

**Data Collation and Truncation Handling:**

* Removed the `_truncate_inputs` method and all related calls, shifting responsibility for truncation entirely to the data collator. Now, if a custom data collator is provided, it must handle truncation before padding.
* Updated the documentation for the `data_collator` argument to clearly state that custom collators must truncate sequences before padding, as the trainer will not apply truncation after collation.

**Code Simplification:**

* Removed the unused `flush_right` import and related logic, further simplifying the codebase.

**Model Input Handling:**

* Updated how model input arguments are constructed in `compute_ref_log_probs` and `_compute_loss`, now directly including all relevant keys from the input dictionary without truncation logic.

---

> [!NOTE]
> **Medium Risk**
> Removes a silent safety net that truncated/padded batches inside `DPOTrainer`, so custom `data_collator`s that relied on that behavior may now produce overlong or misaligned tensors and fail at runtime.
> 
> **Overview**
> **Removes post-collation truncation from `DPOTrainer`.** The internal `_truncate_inputs` path (including `keep_end` flush/realign logic) is deleted, and both `compute_ref_log_probs` and loss computation now consume collator outputs as-is.
> 
> **Updates the trainer/collator contract.** The `data_collator` docstring now explicitly requires custom collators to truncate sequences *before* padding, and model kwargs assembly is simplified to pass through optional fields (e.g., `token_type_ids`, multimodal image inputs) directly without length fix-ups.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 298342251668227069207305a06c6a557414df3d. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/trainer/dpo_trainer.py [hunk 1]
```diff
@@ -446,7 +445,8 @@ class DPOTrainer(_BaseTrainer):
         data_collator ([`~transformers.DataCollator`], *optional*):
             Function to use to form a batch from a list of elements of the processed `train_dataset` or `eval_dataset`.
             Will default to [`~trainer.dpo_trainer.DataCollatorForPreference`] if the model is a language model and
-            [`~trainer.dpo_trainer.DataCollatorForVisionPreference`] if the model is a vision-language model.
+            [`~trainer.dpo_trainer.DataCollatorForVisionPreference`] if the model is a vision-language model. Custom
+            collators must truncate sequences before padding; the trainer does not apply post-collation truncation.
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/trainer/dpo_trainer.py:449]: Removed it.

File: trl/trainer/dpo_trainer.py [hunk 2]
```diff
@@ -446,7 +445,8 @@ class DPOTrainer(_BaseTrainer):
         data_collator ([`~transformers.DataCollator`], *optional*):
             Function to use to form a batch from a list of elements of the processed `train_dataset` or `eval_dataset`.
             Will default to [`~trainer.dpo_trainer.DataCollatorForPreference`] if the model is a language model and
-            [`~trainer.dpo_trainer.DataCollatorForVisionPreference`] if the model is a vision-language model.
+            [`~trainer.dpo_trainer.DataCollatorForVisionPreference`] if the model is a vision-language model. Custom
```
Comment (qgallouedec) [inline trl/trainer/dpo_trainer.py:448]: can you please add the same comment in SFTTrainer.data_collator

**Comment by qgallouedec** [comment]
thanks! I think we might be able to remove `flush_right` from this repo in a next PR

**Comment by qgallouedec** [comment]
@codex review

File: trl/trainer/dpo_trainer.py [hunk 3]
```diff
@@ -1185,16 +1149,15 @@ def _compute_loss(self, model, inputs, return_outputs):
         input_ids = inputs["input_ids"]
         attention_mask = inputs["attention_mask"]
         completion_mask = inputs["completion_mask"]
-        # token_type_ids and mm_token_type_ids are sequence-length-aligned: truncate to match input_ids
-        extra_keys = [k for k in ("token_type_ids", "mm_token_type_ids") if k in inputs]
-        input_ids, attention_mask, completion_mask, *extra = self._truncate_inputs(
-            input_ids, attention_mask, completion_mask, *[inputs[k] for k in extra_keys]
-        )
-
         model_kwargs = {"input_ids": input_ids, "attention_mask": attention_mask, "use_cache": False}
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/dpo_trainer.py:1152]: In my opinion it's actually a good point

File: trl/trainer/dpo_trainer.py (hunk 2 above)
Comment (albertvillanova) [reply trl/trainer/dpo_trainer.py:448]: I am planning to add the same comment to SFT in my subsequent PR, when I remove post-collation truncation from SFT as well.
